### PR TITLE
Add cwd functionality

### LINKED ISSUE #12: cd and relative path binaries
These go together because we currently have no CWD (current working directory) implementation, and relative paths require CWDs (optimally with a way to change it without `exit`ing).

### PR DESCRIPTION
**Pull request by nearlySplat**
Resolves #12

### CODE REVIEW

**Comment by Vendicated** [comment]
Is there a reason you implement this yourself instead of just using process.chdir()? it supports absolute or relative parts out of the box, we'd only need to keep track of the last directory to allow `cd -` 
Screenshot_20210503-211756_Termux

File: src/inbuilts/cd.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+export default class Cd extends InbuiltCommand {
+  public readonly usage = "cd FOLDER";
+  public invoke() {
+    let path = this.args.length ? this.args[0] : (process.env.HOME as string);
+    const [cwd, setCwd, oldCwd] = useCwd();
+    if (path === "-") {
+      path = oldCwd;
+    } else if (!path.startsWith("/")) {
+      path = `${cwd  }/${  path}`;
```
Comment (Vendicated) [inline src/inbuilts/cd.ts:32]: you added a windows check somewhere else so you should add it here too, or rather use path.join()

File: src/parsers/parseArgs.ts [hunk 2]
```diff
@@ -42,8 +44,12 @@ export function parseArgs(str: string) {
     args.push(currentArg);
     currentArg = "";
   }
+  function substitute(char: string) {
+    if (!insideSingleQuotes && !insideDoubleQuotes) currentArg += char;
+  }
 
-  loop: for (const char of str) {
+  loop: for (let i = 0; i < str.length; i++) {
```
Comment (Vendicated) [inline src/parsers/parseArgs.ts:51]: why the conversion to a cstyle for loop?

File: src/util/session.ts [hunk 3]
```diff
@@ -23,6 +23,10 @@ export function prompt(hook: (prompt: string) => void, context = "") {
   hook(`${context || ""}$> `);
 }
 
+export function shortenPath(path: string) {
+  return path.replace(new RegExp(process.env.HOME as string), "~").replace(/(\/)(.).*?(\/)/g, "$1$2$3");
```
Comment (Vendicated) [inline src/util/session.ts:27]: we should avoid using regex if possible as its very slow

File: src/util/session.ts [hunk 4]
```diff
@@ -32,9 +36,9 @@ export function resolveVariable(key: string) {
 /**
  * Prints formatted text to stdout, supports placeholders and automatically stringifies objects
  * @param str Format string
- * @param args Format args. If str has no format placeholders these get appended to str separated by newlines
+ * @param [args] Format args. If str has no format placeholders these get appended to str separated by newlines
  */
-export function printf(str: string, ...args: unknown[]) {
+export function printf<T>(str: string, ...args: T[]) {
```
Comment (Vendicated) [inline src/util/session.ts:41]: what is this for? this doesn't allow passing in mixed types, or does it?

Comment (nearlySplat) [reply src/util/session.ts:41]: it does

File: src/parsers/parseArgs.ts (hunk 2 above)
Comment (nearlySplat) [reply src/parsers/parseArgs.ts:51]: An oversight, I thought I was going to se the index.

File: src/inbuilts/cd.ts (hunk 1 above)
Comment (nearlySplat) [reply src/inbuilts/cd.ts]: fixed

**Comment by Vendicated** [comment]
again why not just use process.chdir()? more straight forward and easier to maintain

**Comment by nearlySplat** [comment]
I thought there might be a reason in the future to preserve the starting current working directory

**Pull request merged**